README.md
# 🌲 Timber.io Logstash Output Plugin

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-output-timber.svg)](https://travis-ci.org/logstash-plugins/logstash-output-timber)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-output-timber.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 1.0.3
  - Docs: Set the default_codec doc attribute.

## 1.0.3
  - Update gemspec summary

## 1.0.3
 - Intelligently map logstash events to the Timber.io log event JSON schema
   (https://github.com/timberio/log-event-json-schema).

## 1.0.2
 - Add hostname to the payload being sent to Timber.

## 1.0.1
 - Drop the @version field since it is not useful to external systems

## 1.0.0
 - Initial release


spec/outputs/timber_spec.rb
require "json"
require "thread"

require "logstash/devutils/rspec/spec_helper"
require "logstash/outputs/timber"
require "logstash/codecs/plain"
require "sinatra"

PORT = rand(65535-1024) + 1025

class LogStash::Outputs::Timber
  attr_writer :agent
  attr_reader :request_tokens
end

# note that Sinatra startup and shutdown messages are directly logged to stderr so
# it is not really possible to disable them without reopening stderr which is not advisable.
#
# == Sinatra (v1.4.6) has taken the stage on 51572 for development with backup from WEBrick
# == Sinatra has ended his set (crowd applauds)

class TestApp < Sinatra::Base
  # disable WEBrick logging
  def self.server_settings
    { :AccessLog => [], :Logger => WEBrick::BasicLog::new(nil, WEBrick::BasicLog::FATAL) }
  end

  def self.add_request(request)
    self.requests << request
  end

  def self.requests
    @requests ||= []
  end

  def self.reset_requests
    @requests = []
  end

  post "/good" do
    self.class.add_request(request)
    [200, "Good"]
  end

  post "/auth_error" do
    self.class.add_request(request)
    [403, "Bad"]
  end

  post "/server_error" do
    self.class.add_request(request)
    [500, "Bad"]
  end
end

RSpec.configure do |config|
  #http://stackoverflow.com/questions/6557079/start-and-call-ruby-http-server-in-the-same-script
  def sinatra_run_wait(app, opts)
    queue = Queue.new

    t = java.lang.Thread.new(
      proc do
        begin
          app.run!(opts) do |server|
            queue.push("started")
          end
        rescue => e
          puts "Error in webserver thread #{e}"
          # ignore
        end
      end
    )
    t.daemon = true
    t.start
    queue.pop # blocks until the run! callback runs
  end

  config.before(:suite) do
    sinatra_run_wait(TestApp, :port => PORT, :server => 'webrick')
    puts "Test webserver on port #{PORT}"
  end
end

describe LogStash::Outputs::Timber do
  let(:port) { PORT }
  let(:output) { described_class.new({"api_key" => "123:abcd1234", "pool_max" => 1}) }
  let(:event) { LogStash::Event.new({"message" => "hi"}) }
  let(:requests) { TestApp.requests }

  before(:each) do
    output.register
    TestApp.reset_requests
  end

  describe "#send_events" do
    it "returns false when the max attempts are exceeded" do
      result = output.send(:send_events, [event], 6)
      expect(result).to eq(false)
    end

    it "returns false when the status is 403" do
      output.url = "http://localhost:#{port}/auth_error"
      result = output.send(:send_events, [event], 1)
      expect(result).to eq(false)
      expect(requests.length).to eq(1)
    end

    it "returns false when the status is 500" do
      allow(output).to receive(:sleep_for_attempt).and_return(0)
      output.url = "http://localhost:#{port}/server_error"
      result = output.send(:send_events, [event], 1)
      expect(result).to eq(false)
      expect(requests.length).to eq(3)
    end

    it "handles fatal request errors" do
      allow(output.send(:http_client)).to receive(:post).and_raise("boom")

      output.url = "http://localhost:#{port}/good"
      result = output.send(:send_events, [event], 1)
      expect(result).to eq(false)
    end

    it "handles retryable request errors" do
      expect(output.send(:http_client)).to receive(:post).exactly(3).times.and_raise(::Manticore::Timeout.new)

      output.url = "http://localhost:#{port}/good"
      result = output.send(:send_events, [event], 1)
      expect(result).to eq(false)
    end

    it "returns true when the status is 200" do
      output.url = "http://localhost:#{port}/good"
      result = output.send(:send_events, [event], 1)
      expect(result).to eq(true)
      expect(requests.length).to eq(1)

      request = requests.first
      expect(request.env["CONTENT_TYPE"]).to eq("application/json")
      expect(request.env["HTTP_AUTHORIZATION"]).to eq("Basic MTIzOmFiY2QxMjM0")
      expect(request.env["HTTP_USER_AGENT"]).to eq("Timber Logstash/1.0.2")

      parsed_body = JSON.parse!(request.body.read)
      expect(parsed_body.length).to eq(1)

      body_event = parsed_body.first
      timestamp_iso8601 = event.get("@timestamp").to_iso8601
      expected_payload = {"$schema"=>"https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json", "dt"=>timestamp_iso8601, "message"=>"hi"}
      expect(body_event).to eq(expected_payload)
    end
  end

  describe "#event_hash" do
    it "merges the timber key" do
      event = LogStash::Event.new({"message" => "my message", "timber" => {"context" => {"system" => {"pid" => 123}}}})
      hash = output.send(:event_hash, event)

      dt = event.get("@timestamp").utc.to_iso8601
      expect(hash).to eq({
        "$schema" => "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json",
        "context" => {"system" => {"pid" => 123}},
        "message" => "my message",
        "dt" => dt
      })
    end

    it "moves host" do
      event = LogStash::Event.new({"message" => "my message", "host" => "local.myhost.com"})
      hash = output.send(:event_hash, event)

      dt = event.get("@timestamp").utc.to_iso8601
      expect(hash).to eq({
        "$schema" => "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json",
        "message" => "my message",
        "dt" => dt,
        "context" => {"system" => {"hostname" => "local.myhost.com"}}
      })
    end

    it "moves everything else to meta" do
      event = LogStash::Event.new({"message" => "my message", "key" => "val"})
      hash = output.send(:event_hash, event)

      dt = event.get("@timestamp").utc.to_iso8601
      expect(hash).to eq({
        "$schema" => "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json",
        "message" => "my message",
        "dt" => dt,
        "meta" => {"key" => "val"}
      })
    end

    it "handles bother timber and host" do
      event = LogStash::Event.new({"message" => "my message", "host" => "local.myhost.com", "timber" => {"context" => {"system" => {"pid" => 123}}}})
      hash = output.send(:event_hash, event)

      dt = event.get("@timestamp").utc.to_iso8601
      expect(hash).to eq({
        "$schema" => "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json",
        "message" => "my message",
        "dt" => dt,
        "context" => {"system" => {"hostname" => "local.myhost.com", "pid" => 123}}
      })
    end
  end
end


lib/logstash/outputs/timber/http_client.rb
# encoding: utf-8
require "logstash/config/mixin"
require "manticore"

# This contains the HTTP client code within it's own namespace. It is based off
# of the logstash-mixin-http_client plugin.
class LogStash::Outputs::Timber < LogStash::Outputs::Base
  module HttpClient
    class InvalidHTTPConfigError < StandardError; end

    def self.included(base)
      base.extend(self)
      base.setup_http_client_config
    end

    def setup_http_client_config
      # Timeout (in seconds) for the entire request
      config :request_timeout, :validate => :number, :default => 60

      # Timeout (in seconds) to wait for data on the socket. Default is `10s`
      config :socket_timeout, :validate => :number, :default => 10

      # Timeout (in seconds) to wait for a connection to be established. Default is `10s`
      config :connect_timeout, :validate => :number, :default => 10

      # Max number of concurrent connections. Defaults to `50`
      config :pool_max, :validate => :number, :default => 50

      # If you need to use a custom X.509 CA (.pem certs) specify the path to that here
      config :cacert, :validate => :path

      # If you'd like to use a client certificate (note, most people don't want this) set the path to the x509 cert here
      config :client_cert, :validate => :path

      # If you're using a client certificate specify the path to the encryption key here
      config :client_key, :validate => :path

      # If you need to use a custom keystore (`.jks`) specify that here. This does not work with .pem keys!
      config :keystore, :validate => :path

      # Specify the keystore password here.
      # Note, most .jks files created with keytool require a password!
      config :keystore_password, :validate => :password

      # Specify the keystore type here. One of `JKS` or `PKCS12`. Default is `JKS`
      config :keystore_type, :validate => :string, :default => "JKS"

      # If you need to use a custom truststore (`.jks`) specify that here. This does not work with .pem certs!
      config :truststore, :validate => :path

      # Specify the truststore password here.
      # Note, most .jks files created with keytool require a password!
      config :truststore_password, :validate => :password

      # Specify the truststore type here. One of `JKS` or `PKCS12`. Default is `JKS`
      config :truststore_type, :validate => :string, :default => "JKS"

      # If you'd like to use an HTTP proxy . This supports multiple configuration syntaxes:
      #
      # 1. Proxy host in form: `http://proxy.org:1234`
      # 2. Proxy host in form: `{host => "proxy.org", port => 80, scheme => 'http', user => 'username@host', password => 'password'}`
      # 3. Proxy host in form: `{url =>  'http://proxy.org:1234', user => 'username@host', password => 'password'}`
      config :proxy
    end

    def build_http_client_config
      c = {
        connect_timeout: @connect_timeout,
        socket_timeout: @socket_timeout,
        request_timeout: @request_timeout,
        follow_redirects: true,
        automatic_retries: 1,
        retry_non_idempotent: true,
        check_connection_timeout: 200,
        pool_max: @pool_max,
        pool_max_per_route: @pool_max,
        cookies: false,
        keepalive: true
      }

      if @proxy
        # Symbolize keys if necessary
        c[:proxy] = @proxy.is_a?(Hash) ?
          @proxy.reduce({}) {|memo,(k,v)| memo[k.to_sym] = v; memo} :
          @proxy
      end

      c[:ssl] = {}
      if @cacert
        c[:ssl][:ca_file] = @cacert
      end

      if @truststore
        c[:ssl].merge!(
          :truststore => @truststore,
          :truststore_type => @truststore_type,
          :truststore_password => @truststore_password.value
        )

        if c[:ssl][:truststore_password].nil?
          raise LogStash::ConfigurationError, "Truststore declared without a password! This is not valid, please set the 'truststore_password' option"
        end
      end

      if @keystore
        c[:ssl].merge!(
          :keystore => @keystore,
          :keystore_type => @keystore_type,
          :keystore_password => @keystore_password.value
        )

        if c[:ssl][:keystore_password].nil?
          raise LogStash::ConfigurationError, "Keystore declared without a password! This is not valid, please set the 'keystore_password' option"
        end
      end

      if @client_cert && @client_key
        c[:ssl][:client_cert] = @client_cert
        c[:ssl][:client_key] = @client_key
      elsif !!@client_cert ^ !!@client_key
        raise InvalidHTTPConfigError, "You must specify both client_cert and client_key for an HTTP client, or neither!"
      end

      c
    end

    def http_client
      @http_client ||= Manticore::Client.new(build_http_client_config)
    end
  end
end

lib/logstash/outputs/timber.rb
# encoding: utf-8

# Core classes
require "base64"
require "uri"

# Logstash classes
require "logstash/namespace"
require "logstash/json"
require "logstash/outputs/base"
require "logstash/outputs/timber/http_client"

# This is an output class that intelligently forwards logstash events to the Timber.io service.
#
# For a comprehensive overview around how this works and the various configuration options,
# please see: https://timber.io/docs/platforms/logstash
class LogStash::Outputs::Timber < LogStash::Outputs::Base
  include HttpClient

  VERSION = "1.0.2".freeze
  CONTENT_TYPE = "application/json".freeze
  JSON_SCHEMA = "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json".freeze
  MAX_ATTEMPTS = 3
  METHOD = :post.freeze
  RETRYABLE_MANTICORE_EXCEPTIONS = [
    ::Manticore::Timeout,
    ::Manticore::SocketException,
    ::Manticore::ClientProtocolException,
    ::Manticore::ResolutionFailure,
    ::Manticore::SocketTimeout
  ].freeze
  RETRYABLE_CODES = [429, 500, 502, 503, 504].freeze
  URL = "https://logs.timber.io/frames".freeze
  USER_AGENT = "Timber Logstash/#{VERSION}".freeze

  # Attribute for testing purposes only
  attr_writer :url

  concurrency :shared

  # This output lets you send events to the Timber.io logging service.
  #
  # This output will execute up to 'pool_max' requests in parallel for performance.
  # Consider this when tuning this plugin for performance.
  #
  # Additionally, note that when parallel execution is used strict ordering of events is not
  # guaranteed!
  #
  # Beware, this gem does not yet support codecs. Please use the 'format' option for now.
  config_name "timber"

  # Your Timber API key. This can be obtained by creating an app at https://app.timber.io.
  # Already have an app? You can find your API key in your app's settings.
  config :api_key, :validate => :string, :required => :true

  def register
    encoded_api_key = Base64.urlsafe_encode64(@api_key).chomp
    authorization_value = "Basic #{encoded_api_key}"
    @headers = {
      "Authorization" => authorization_value,
      "Content-Type" => CONTENT_TYPE,
      "User-Agent" => USER_AGENT
    }
    @url = URL
  end

  def multi_receive(events)
    send_events(events, 1)
  end

  def close
    http_client.close
  end

  private
    def send_events(events, attempt)
      if attempt > MAX_ATTEMPTS
        @logger.warn(
          "Max attempts exceeded, dropping events",
          :attempt => attempt
        )
        return false
      end

      response =
        begin
          hash_events = events.collect { |e| event_hash(e) }
          body = LogStash::Json.dump(hash_events)
          http_client.post(@url, :body => body, :headers => @headers)
        rescue Exception => e
          if retryable_exception?(e)
            @logger.warn(
              "Attempt #{attempt}, retryable exception when making request",
              :attempt => attempt,
              :class => e.class.name,
              :message => e.message,
              :backtrace => e.backtrace
            )
            return send_events(events, attempt + 1)
          else
            @logger.error(
              "Attempt #{attempt}, fatal exception when making request",
              :attempt => attempt,
              :class => e.class.name,
              :message => e.message,
              :backtrace => e.backtrace
            )
            return false
          end
        end

      code = response.code

      if code >= 200 && code <= 299
        true
      elsif RETRYABLE_CODES.include?(code)
        @logger.warn(
          "Bad retryable response from the Timber API",
          :attempt => attempt,
          :code => code
        )
        sleep_time = sleep_for_attempt(attempt)
        sleep(sleep_time)
        send_events(events, attempt + 1)
      else
        @logger.error(
          "Bad fatal response from the Timber API",
          :attempt => attempt,
          :code => code
        )
        false
      end
    end

    # This method takes a `Logstash::Event` object and converts it into a hash
    # that is acceptable by the Timber API. Each event is converted into a JSON
    # document that conforms to the Timber log event JSON schema:
    #
    # https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.1.1/schema.json
    #
    # This realized by the following steps:
    #
    # 1. Timber will look for specific keys and map them to the appropriate keys as defined
    #    in our log event JSON schema. Specifically `@timestamp`, `host`, and `message`.
    # 2. If a `timber` key is present it _must_ be a hash that conforms to the Timber log event
    #    JSON schema. This hash will be merged in before being sent to Timber.
    # 3. All other root level keys will be treated as generic JSON and will be made available
    #    in Timber as they would kibana, etc.
    def event_hash(e)
      timber_hash = {"$schema" => JSON_SCHEMA}
      event_hash = e.to_hash

      # Delete unused logstash specific attributes
      event_hash.delete("@version")

      # Map the timber key first since we merge in values
      # later.
      timber = event_hash.delete("timber")
      if timber.is_a?(Hash)
        timber_hash.merge!(timber)
      end

      # Map the timestamp
      timestamp = event_hash.delete("@timestamp")

      if timestamp
        timber_hash["dt"] ||= timestamp.utc.to_iso8601
      end

      # Map the host
      host = event_hash.delete("host")

      if host
        timber_hash["context"] ||= {}
        timber_hash["context"]["system"] ||= {}
        timber_hash["context"]["system"]["hostname"] ||= host
      end

      # Map the message
      message = event_hash.delete("message")

      if message
        timber_hash["message"] ||= message
      end

      # Move everything else to meta, merging to preseve previous meta values.
      if event_hash != {}
        timber_hash["meta"] ||= {}
        timber_hash["meta"].merge!(event_hash)
      end

      timber_hash
    end

    def retryable_exception?(e)
      RETRYABLE_MANTICORE_EXCEPTIONS.any? do |exception_class|
        e.is_a?(exception_class)
      end
    end

    def sleep_for_attempt(attempt)
      sleep_for = attempt ** 2
      sleep_for = sleep_for <= 60 ? sleep_for : 60
      (sleep_for / 2) + (rand(0..sleep_for) / 2)
    end
end
